### ISSUE TITLE
How can I access GPIOs

### ISSUE THREAD

**Issue by peerst** [grisp/grisp2-rtems-toolchain#21]
First use case: accessing the LEDs without using grisp_led functions

**Comment by c-mauderer** [comment]
I added a simple GPIO API to the BSP (see https://github.com/grisp/rtems/blob/grisp2-devel-20200603/bsps/arm/imx/include/bsp/imx-gpio.h). If you want to use that:

You can either search for a GPIO via the FDT like I did that for the LEDs and initialize the GPIO structures from that:
https://github.com/grisp/libgrisp/blob/master/src/led.c#L136

Alternatively you could initialize the structures by hand and use them via the API. An sample (written from head) for some random Pin GPIO2_10 (no idea what that pin does):
```
#include <bsp/imx-gpio.h>
/* ... */
struct imx_gpio_pin some_pin = {
    .gpio = IMX_GPIO2,
    .mask = 1 << 10,
    .shift = 10,
    .mode = IMX_GPIO_MODE_OUTPUT,
};
imx_gpio_init(&some_pin);  /* Initializes the registers with direction and similar */
while (true) {
    imx_gpio_toggle_output(&some_pin);
    sleep(1);
}
```

As an alternative you should be able to use the FreeBSD API. There is an adaption layer in libbsd to allow usage of GPIO pins for some drivers (for example switching USB supply on). But I never tried how to use it with an user space API.

**Comment by peerst** [comment]
So this is handled completely different from GRiSP 1 ?

I thought GRiSP 1 was using a GPIO HAL approach but apparently this is different for each BSP

Kind of weird architecture that we have portable driver lib calls for many things like I2C and SPI but GPIO is BSP dependent.

**Comment by c-mauderer** [comment]
Yes, there is no (usable) uniform GPIO API in RTEMS and most BSPs do it different.

**Comment by c-mauderer** [comment]
I think for GRiSP1 it was a generic API that has been provided by Atmel for the ATSAM. That library (libchip) was just imported into the BSP.

It's definitively something that is missing in RTEMS. No one had the time / budget / pressure to suggest, discuss and implement a good uniform API for GPIO yet.

**Comment by peerst** [comment]
No problem for us since we have a kind of Erlang+C level BSP structure which allows different drivers for different architectures

Was just confused

**Comment by eproxus** [comment]
Solved and will be refined with #33.
